source/common/network/listen_socket_impl.h

```diff
@@ -43,18 +44,19 @@ class SocketImpl : public virtual Socket {
   const OptionsSharedPtr& options() const override { return options_; }
 
 protected:
-  SocketImpl(int fd, const Address::InstanceConstSharedPtr& local_address)
-      : fd_(fd), local_address_(local_address) {}
+  SocketImpl(const IoHandlePtr& io_handle, const Address::InstanceConstSharedPtr& local_address)
```

Comment:
I think we need to figure out who is the end-owner of the io_handle, and make ownership transfer explicit.
For example, If the SocketImpl is taking ownership, this should be an explicit && and should not be taking in a const IoHandlePtr.
Creating and destroying IoHandles may work fine for socketIoHandles where we're basically adding a shim wrapper around an fd, but it's not necessarily going to work for something like VPP or QUIC where there may be significant state in the underlying class and passing in a cosnt IoHandlePtr but the using it to make a non-const IOHandle using the same fd is a bit misleading.

Reply:
I feel SocketImpl is a good place to own the IoHandle instance. If so, Socket::ioHandle() should return a reference or a pointer instead of passing ownership around.

I'm not familiar with Envoy code style, is reference preferred over pointer?

Reply:
I'm confused here- does it matter who owns it if IoHandlePtr is a shared pointer? Am I missing your point?

Reply:
shared_ptr is more expensive than unique_ptr. And I *think* we don't have to use shared_ptr here.